File: internal/server/events.go
package server

import (
	"time"

	"github.com/sourcehawk/triagent/internal/claude"
)

// EventEnvelope is the JSON shape we ship over SSE for one transcript
// event. The Kind field discriminates the optional fields. Seq is a
// monotonic per-investigation counter that doubles as the SSE message id,
// so a reconnecting client can resume via Last-Event-ID without dupes.
type EventEnvelope struct {
	Seq       int       `json:"seq"`
	Kind      string    `json:"kind"`
	Timestamp time.Time `json:"timestamp"`

	// Per-kind optional fields. JSON-omitempty so the wire is compact.
	SessionID    string         `json:"sessionId,omitempty"`
	Subtype      string         `json:"subtype,omitempty"`
	Text         string         `json:"text,omitempty"`
	ToolID       string         `json:"toolId,omitempty"`
	ToolName     string         `json:"toolName,omitempty"`
	ToolInput    map[string]any `json:"toolInput,omitempty"`
	ParentToolID string         `json:"parentToolId,omitempty"`
	Origin       string         `json:"origin,omitempty"`

	// Set on push_state envelopes only. Carries the new state of an
	// async session push (started, succeeded, failed). The frontend
	// uses these to update the SessionView and the toast.
	PushState *PushStatePayload `json:"pushState,omitempty"`

	// Set on rehydrate_state envelopes only. Carries the phase of the
	// rehydrate orchestration (started / succeeded / failed) so the
	// frontend can render a banner without polling.
	RehydrateState *RehydrateStatePayload `json:"rehydrateState,omitempty"`

	// Set on auto_mode_state envelopes only. Carries the phase of the
	// auto-operator (started / paused / resumed / finished / aborted) and
	// (when Warning is set) an advisory-only annotation.
	AutoMode *AutoModePayload `json:"autoMode,omitempty"`

	// Usage is the message-level (assistant) or invocation-level (result)
	// token tally claude reports on the wire. nil for synthetic envelopes
	// and for claude events that omit the block. Frontend sums Usage on
	// result envelopes to compute the session total.
	Usage *claude.Usage `json:"usage,omitempty"`
	// CostUSD rides on result envelopes only. Zero elsewhere. Frontend
	// sums across result envelopes to compute the session cost.
	CostUSD float64 `json:"costUsd,omitempty"`
}

// PushStatePayload rides on a push_state envelope. Phase is always set;
// the rest depend on the phase. "pushing" carries StartedAt; "success"
// carries URL/Branch/Base; "error" carries Error.
type PushStatePayload struct {
	Phase     string     `json:"phase"`
	StartedAt *time.Time `json:"startedAt,omitempty"`
	URL       string     `json:"url,omitempty"`
	Branch    string     `json:"branch,omitempty"`
	Base      string     `json:"base,omitempty"`
	Error     string     `json:"error,omitempty"`
}

// RehydrateStatePayload rides on a rehydrate_state envelope. Phase is
// always set ("started", "succeeded", "failed"). Error is set only for
// "failed". Degraded carries human-readable downgrades the orchestrator
// applied (e.g. "prom" when the prometheus port-forward couldn't come
// up); empty for fully-successful rehydrates.
type RehydrateStatePayload struct {
	Phase    string   `json:"phase"`
	Error    string   `json:"error,omitempty"`
	Degraded []string `json:"degraded,omitempty"`
}

// AutoModePayload rides on an auto_mode_state envelope.
type AutoModePayload struct {
	Phase       string `json:"phase"` // started | paused | resumed | finished | aborted
	Reason      string `json:"reason,omitempty"`
	TakenOverBy string `json:"takenOverBy,omitempty"` // "human" | "operator"
	Error       string `json:"error,omitempty"`       // populated on aborted phase, or when Warning is set, to describe the issue.
	Warning     bool   `json:"warning,omitempty"`     // true means the envelope is advisory only and the phase did not change.
}

// Envelope kinds. The "user" and "end" kinds are synthesised by the
// server (they don't come from claude); the rest mirror claude.Event.
const (
	envKindSystem     = "system"
	envKindAssistant  = "assistant"
	envKindToolUse    = "tool_use"
	envKindToolResult = "tool_result"
	envKindToolStatus = "tool_status" // synthetic: a sub-agent emitted a STATUS marker
	envKindResult     = "result"
	envKindError      = "error"
	envKindUser       = "user"       // synthetic: a follow-up prompt the operator submitted
	envKindEnd        = "end"        // synthetic: claude exited; transcript is now idle
	envKindPushState      = "push_state"      // synthetic: an async push transitioned phase
	envKindLabel          = "label"           // synthetic: Investigation.Label was changed
	envKindRehydrateState = "rehydrate_state" // synthetic: rehydrate orchestration phase changed
	envKindAutoModeState  = "auto_mode_state" // synthetic: auto-operator phase transition (started/paused/resumed/finished/aborted)
	envKindUsage          = "usage"           // accounting-only: per-API-call token tally for the running totals
)

// fromClaudeEvent converts a claude.Event into an EventEnvelope. Returns
// (env, true) if the event has anything worth sending, (zero, false) for
// silent events (Unknown, empty assistant text, tool_*).
//
// Tool calls are intentionally dropped here. Claude's stream-json output
// doesn't reliably emit tool_use / tool_result for every call (they live
// inside assistant/user content blocks and the wire shape varies across
// versions), so we treat the stream as advisory for tool activity and
// rely on triagent-mcp's POSTed telemetry as the source of truth instead. That
// telemetry path lands in publishToolEvent below.
func fromClaudeEvent(e claude.Event) (EventEnvelope, bool) {
	env := EventEnvelope{}
	switch e.Kind {
	case claude.EventSystem:
		env.Kind = envKindSystem
		env.Subtype = e.Subtype
		env.SessionID = e.SessionID
	case claude.EventAssistant:
		if e.Text == "" {
			// Empty-text assistants (the synthetic fallback for messages
			// that were pure tool_use) carry no UI payload. Usage now flows
			// through a dedicated EventUsage event in the same parseLine
			// slice, so dropping this here loses nothing.
			return EventEnvelope{}, false
		}
		env.Kind = envKindAssistant
		env.Text = e.Text
		env.Usage = e.Usage
	case claude.EventResult:
		env.Kind = envKindResult
		env.Subtype = e.Subtype
		env.Text = e.FinalText
		// e.Usage on a result line is the LAST API call's snapshot, not the
		// invocation total — claude reports it that way. We deliberately
		// drop it here and rely on the per-call EventUsage events for token
		// totals. CostUSD is the invocation roll-up and still flows.
		env.CostUSD = e.CostUSD
	case claude.EventUsage:
		env.Kind = envKindUsage
		env.Usage = e.Usage
	case claude.EventError:
		env.Kind = envKindError
		if e.Err != nil {
			env.Text = e.Err.Error()
		}
	case claude.EventToolUse, claude.EventToolResult, claude.EventUnknown:
		return EventEnvelope{}, false
	default:
		return EventEnvelope{}, false
	}
	return env, true
}

File: frontend/lib/events.ts
// EventEnvelope mirrors the JSON shape the Go server emits over SSE.
// Optional fields are present per kind; the reducer below folds a stream
// of envelopes into a structured transcript the UI can render directly.

import type { RehydrateStatePayload } from "@/lib/api";

// AutoModePhase enumerates the auto-operator lifecycle states the Go
// server emits on auto_mode_state envelopes. Kept narrow on purpose —
// matches the AutoModePayload.Phase comment in
// investigate/internal/server/events.go.
export type AutoModePhase =
  | "started"
  | "paused"
  | "resumed"
  | "finished"
  | "aborted";

// AutoModePayload mirrors the Go AutoModePayload struct
// (investigate/internal/server/events.go). Rides on auto_mode_state
// envelopes. The useAutoMode hook folds these into the UI's auto-mode
// state; the transcript reducer ignores them.
export type AutoModePayload = {
  phase: AutoModePhase;
  reason?: string;
  takenOverBy?: "human" | "operator";
  error?: string;
  warning?: boolean;
};

export type EventEnvelope = {
  seq: number;
  kind:
    | "system"
    | "assistant"
    | "tool_use"
    | "tool_result"
    | "tool_status"
    | "result"
    | "error"
    | "user"
    | "end"
    | "push_state"
    | "label"
    | "rehydrate_state"
    | "codefix_pr_state"
    | "auto_mode_state"
    | "usage";
  timestamp: string;
  sessionId?: string;
  subtype?: string;
  text?: string;
  toolId?: string;
  toolName?: string;
  toolInput?: Record<string, unknown>;
  parentToolId?: string;
  // origin labels who authored a user envelope. "human" → the chat
  // composer; "operator" → the auto-operator's send_message tool. The
  // server stamps this on the EventEnvelope so the transcript can style
  // operator-authored user turns distinctly (pink envelope). Absent on
  // legacy user envelopes; treat undefined as "human".
  origin?: "human" | "operator";
  // Set on push_state envelopes only. The server emits these as the
  // async session-push goroutine transitions phase.
  pushState?: {
    phase: "pushing" | "success" | "error";
    startedAt?: string;
    url?: string;
    branch?: string;
    base?: string;
    error?: string;
  };
  // Set on rehydrate_state envelopes only. The server emits these as
  // the rehydrate orchestration transitions phase (started →
  // succeeded | failed).
  rehydrateState?: RehydrateStatePayload;
  // codefixPRState is set on codefix_pr_state envelopes. The launcher's
  // PR-state poller emits one whenever a tracked codefix PR transitions
  // state (open → merged/closed). The streaming reducer routes it into
  // a per-URL state map; the CodefixProposalCard reads from there.
  codefixPRState?: {
    url: string;
    state: "open" | "merged" | "closed" | "draft" | "";
    mergedAt?: string;
    closedAt?: string;
  };
  // autoMode is set on auto_mode_state envelopes only. The server emits
  // these as the AutoOperator transitions phase (started → paused |
  // resumed | finished | aborted). The useAutoMode hook folds these
  // into UI state; the transcript reducer ignores them.
  autoMode?: AutoModePayload;
  // usage rides on `usage` envelopes (per-API-call token tally, summed
  // across the invocation), on assistant envelopes for the message-level
  // tally on text replies (informational, not summed), and on result
  // envelopes legacy-only (older sessions). Optional because older /
  // unusual claude builds may omit it. The sidebar + session footer sum
  // `usage`-envelope usage to render running totals.
  usage?: Usage;
  // costUsd rides on result envelopes only. Summed across result
  // envelopes for the per-investigation total.
  costUsd?: number;
};

// Usage mirrors the Go claude.Usage struct (investigate/internal/claude/events.go).
// Every component is in tokens; cache_creation_input_tokens and
// cache_read_input_tokens are distinct from input/output so frontends
// that want a cost-aware "in / out / cache" breakdown can have it.
export type Usage = {
  inputTokens?: number;
  outputTokens?: number;
  cacheCreationInputTokens?: number;
  cacheReadInputTokens?: number;
};

// NestedEvent is a child of a tool_call card — surfaces as collapsed rows
// inside the parent. Used to render sub-agent activity (e.g. claude reading
// files inside a git MCP's analyze_change tool) without polluting the main
// transcript.
export type NestedEvent =
  | { kind: "text"; id: string; text: string; ts: string }
  | { kind: "status"; id: string; text: string; ts: string }
  | {
      kind: "tool";
      id: string;
      toolId: string;
      name: string;
      input?: Record<string, unknown>;
      result?: string;
      errorText?: string;
      ts: string;
      tsEnd?: string;
    };

export type TranscriptItem =
  // origin labels who authored a user envelope. Carried through from
  // the wire EventEnvelope so the transcript renderer can style
  // operator-authored user turns distinctly (pink envelope chrome).
  | { kind: "user"; id: string; text: string; origin?: "human" | "operator" }
  | { kind: "assistant"; id: string; text: string }
  | {
      kind: "tool_call";
      id: string;
      toolId: string;
      name: string;
      input?: Record<string, unknown>;
      result?: string;
      startedAt: string; // from tool_use envelope timestamp
      endedAt?: string; // from tool_result envelope timestamp; absent while running
      children?: NestedEvent[];
    }
  | { kind: "error"; id: string; text: string }
  // auto_mode_divider is a synthetic transcript item folded out of
  // auto_mode_state envelopes by applyEvent. Rendered as an inline
  // "— auto mode <phase> —" chip in the transcript so the phase
  // transitions interleave with chat turns in the order they happened.
  // The useAutoMode hook still reads auto_mode_state envelopes for the
  // composer-gating UI state; the divider is purely visual.
  | {
      kind: "auto_mode_divider";
      id: string;
      payload: AutoModePayload;
      timestamp: string;
    };

// Wire-format name of the strategies summarize tool. SessionView
// renders this tool's result inline as the amber conclusion block —
// the markdown audit trail is the conclusion content, no separate
// marker tool needed. Multiple calls (e.g. one per material follow-up)
// each get their own block, traceable in order.
export const SUMMARIZE_TOOL_NAME = "mcp__triagent-strategies__summarize";

// Wire-format name of the playbook-proposal draft tool. SessionView
// renders this tool's result inline as a diff card with Approve /
// Decline controls; the structured-output JSON carries
// {proposal_id, base_yaml, new_yaml, …}. The agent
// is instructed (via the playbook_proposal meta-playbook + system
// prompt) to call this preemptively, no chat-side ask required.
export const PROPOSE_PLAYBOOK_DRAFT_TOOL_NAME =
  "mcp__triagent-strategies__playbook_proposal_draft";

// Wire-format name of the wiki-proposal draft tool. SessionView
// renders this tool's result inline as a WikiProposalCard with
// markdown preview + Approve / Decline controls; the
// structured-output JSON carries
// {kind: "wiki_proposal_draft", proposal_id, slug, base_md,
//  new_md, is_new, ...}.
export const PROPOSE_WIKI_DRAFT_TOOL_NAME =
  "mcp__triagent-wiki__propose_wiki_draft";

// Wire-format suffix of the per-repo draft_pr tool. Since each linked
// repo gets its own triagent-git-<alias> MCP server (one process per repo),
// the full tool name varies (e.g. `mcp__triagent-git-zeebe__draft_pr`,
// `mcp__triagent-git-operate__draft_pr`). isDraftPrToolName matches the
// prefix + suffix shape without enumerating aliases.
export const DRAFT_PR_TOOL_NAME_SUFFIX = "__draft_pr";

export function isDraftPrToolName(name: string): boolean {
  return name.startsWith("mcp__triagent-git-") && name.endsWith(DRAFT_PR_TOOL_NAME_SUFFIX);
}

// Wire-format suffix of the per-repo create_github_issue tool. Like
// draft_pr, the full name varies per linked repo
// (mcp__triagent-git-<alias>__create_github_issue). Mirrors the backend
// isCreateGithubIssueToolName helper in
// internal/server/handlers_codefix.go.
export const CREATE_GITHUB_ISSUE_TOOL_NAME_SUFFIX = "__create_github_issue";

export function isCreateGithubIssueToolName(name: string): boolean {
  return (
    name.startsWith("mcp__triagent-git-") &&
    name.endsWith(CREATE_GITHUB_ISSUE_TOOL_NAME_SUFFIX)
  );
}

// CodefixProposalPayload is the JSON shape the launcher emits on
// GET /api/codefix-proposals/{id} and the result body of a draft_pr
// tool call. The CodefixProposalCard renders this directly.
export type CodefixProposalPayload = {
  proposal_id: string;
  repo: string;            // owner/name
  issue_url: string;
  issue_number: number;
  pr_url: string;
  pr_number: number;
  branch_name: string;
  summary: string;
  // investigation_id and created_at land on entries persisted by
  // newer launchers — the get-by-id handler doesn't depend on them
  // but the list handler relies on both for traceability + sort.
  investigation_id?: string;
  created_at?: string;
};

// CodefixProposalListing is the GET /api/codefix-proposals list-item
// shape: the payload joined with current PR state from the
// launcher's poller cache. The repos sidenav consumes this.
export type CodefixProposalListing = CodefixProposalPayload & {
  pr_state?: string;
};

// applyEvent folds a single envelope into the running transcript. tool_use
// becomes a tool_call card; the matching tool_result fills in `.result`
// in-place. Events with parentToolId attach as nested children of that
// parent tool_call rather than entering the top-level list.
// system + end are intentionally not rendered as transcript items — they
// drive the status pill instead.
export function applyEvent(
  items: TranscriptItem[],
  ev: EventEnvelope,
): TranscriptItem[] {
  // Nested events route to a parent tool_call's `children` array. The
  // parent must already exist (parent tool_use precedes its children in
  // the wire ordering); if it doesn't, we drop the event silently rather
  // than synthesising a top-level item that would confuse the UI.
  if (ev.parentToolId && (ev.kind === "tool_use" || ev.kind === "tool_result" || ev.kind === "assistant" || ev.kind === "tool_status")) {
    return items.map((it) =>
      it.kind === "tool_call" && it.toolId === ev.parentToolId
        ? { ...it, children: applyNested(it.children ?? [], ev) }
        : it,
    );
  }

  switch (ev.kind) {
    case "user":
      return [
        ...items,
        { kind: "user", id: idFor(ev), text: ev.text ?? "", origin: ev.origin },
      ];
    case "assistant":
      return [
        ...items,
        { kind: "assistant", id: idFor(ev), text: ev.text ?? "" },
      ];
    case "tool_use":
      return [
        ...items,
        {
          kind: "tool_call",
          id: idFor(ev),
          toolId: ev.toolId ?? "",
          name: ev.toolName ?? "",
          input: ev.toolInput,
          startedAt: ev.timestamp,
        },
      ];
    case "tool_result":
      return items.map((it) =>
        it.kind === "tool_call" && it.toolId === ev.toolId
          ? { ...it, result: ev.text ?? "", endedAt: ev.timestamp }
          : it,
      );
    case "result":
      // No transcript item: the result envelope used to render an
      // amber summary block, but it fired per-turn — every follow-up
      // got framed as a summary. The formal conclusion is now opt-in
      // via the conclude_investigation marker tool.
      return items;
    case "error":
      return [...items, { kind: "error", id: idFor(ev), text: ev.text ?? "" }];
    case "auto_mode_state":
      // auto_mode_state drives two UI surfaces. The useAutoMode hook
      // folds it into the composer-gating state (Take over / Resume /
      // Restart). It ALSO produces a synthetic transcript item — a
      // small pink-bordered divider chip — so the operator can see in
      // the chat scroll where the auto-operator paused, resumed,
      // finished, or aborted. We render the divider via the existing
      // TranscriptItem pipeline so it interleaves with user/assistant
      // turns by seq order without a parallel render pass.
      if (!ev.autoMode) return items;
      return [
        ...items,
        {
          kind: "auto_mode_divider",
          id: idFor(ev),
          payload: ev.autoMode,
          timestamp: ev.timestamp,
        },
      ];
    case "system":
    case "end":
    case "push_state":
    case "label":
    case "rehydrate_state":
    case "tool_status":
    case "codefix_pr_state":
    case "usage":
      // usage envelopes carry per-API-call token tallies for the
      // running totals — SessionWorkspace folds them into investigation
      // state directly. They are not transcript items.
      return items;
  }
}

function applyNested(children: NestedEvent[], ev: EventEnvelope): NestedEvent[] {
  switch (ev.kind) {
    case "assistant":
      return [
        ...children,
        { kind: "text", id: idFor(ev), text: ev.text ?? "", ts: ev.timestamp },
      ];
    case "tool_status":
      return [
        ...children,
        { kind: "status", id: idFor(ev), text: ev.text ?? "", ts: ev.timestamp },
      ];
    case "tool_use":
      return [
        ...children,
        {
          kind: "tool",
          id: idFor(ev),
          toolId: ev.toolId ?? "",
          name: ev.toolName ?? "",
          input: ev.toolInput,
          ts: ev.timestamp,
        },
      ];
    case "tool_result":
      return children.map((c) =>
        c.kind === "tool" && c.toolId === ev.toolId
          ? { ...c, result: ev.text ?? "", tsEnd: ev.timestamp }
          : c,
      );
    default:
      return children;
  }
}

function idFor(ev: EventEnvelope): string {
  return `${ev.kind}-${ev.seq}`;
}

// reduce folds an entire batch of envelopes into a transcript. Used for
// the SSE backlog replay on connect — cheaper than dispatching N times
// through React's state setter.
export function reduce(events: EventEnvelope[]): TranscriptItem[] {
  return events.reduce<TranscriptItem[]>(applyEvent, []);
}

// SessionStatus is the high-level state shown by the status pill. Driven
// by `streaming` from /api/investigations and the `end` event over SSE.
export type SessionStatus = "starting" | "streaming" | "idle" | "error";

// parseToolName splits the wire-format tool name `mcp__<server>__<tool>`
// into its server alias (triagent-k8s, prom, triagent-strategies, example-docs,
// triagent-git-<alias>) and the short tool name. Tools without the mcp__ prefix
// come back with server=null and the original name as `short`.
export function parseToolName(name: string): { server: string | null; short: string } {
  if (!name.startsWith("mcp__")) return { server: null, short: name };
  const rest = name.slice(5);
  const idx = rest.indexOf("__");
  if (idx < 0) return { server: null, short: rest };
  return { server: rest.slice(0, idx), short: rest.slice(idx + 2) };
}
